components/zcash_protocol/src/value.rs

```diff
@@ -237,36 +238,41 @@ impl Mul<usize> for ZatBalance {
 /// A Zatoshis can only be constructed from an integer that is within the valid monetary
 /// range of `{0..MAX_MONEY}` (where `MAX_MONEY` = 21,000,000 × 10⁸ zatoshis).
 #[derive(Clone, Copy, Debug, PartialEq, PartialOrd, Eq, Ord)]
-pub struct Zatoshis(ZatBalance);
+pub struct Zatoshis(u64);
 
 impl Zatoshis {
     /// Returns the identity `Zatoshis`
-    pub const ZERO: Self = Zatoshis(ZatBalance(0));
+    pub const ZERO: Self = Zatoshis(0);
 
     /// Returns this Zatoshis as a u64.
     pub fn into_u64(self) -> u64 {
-        self.0.try_into().unwrap()
+        self.0
     }
 
     /// Creates a Zatoshis from a u64.
     ///
     /// Returns an error if the amount is outside the range `{0..MAX_MONEY}`.
     pub fn from_u64(amount: u64) -> Result<Self, ()> {
-        ZatBalance::from_u64(amount).map(Zatoshis)
+        if (0..=MAX_MONEY).contains(&amount) {
+            Ok(Zatoshis(amount))
+        } else {
+            Err(())
+        }
     }
 
     /// Creates a constant Zatoshis from a u64.
     ///
-    /// Panics: if the amount is outside the range `{-MAX_MONEY..MAX_MONEY}`.
+    /// Panics: if the amount is outside the range `{0..MAX_MONEY}`.
     pub const fn const_from_u64(amount: u64) -> Self {
-        Zatoshis(ZatBalance::const_from_u64(amount))
+        assert!(amount <= MAX_MONEY); // contains is not const
+        Zatoshis(amount)
     }
 
     /// Creates a Zatoshis from an i64.
     ///
     /// Returns an error if the amount is outside the range `{0..MAX_MONEY}`.
     pub fn from_nonnegative_i64(amount: i64) -> Result<Self, ()> {
-        ZatBalance::from_nonnegative_i64(amount).map(Zatoshis)
+        u64::try_from(amount).map(Zatoshis).map_err(|_| ())
```

Comment:
This should go through `Zatoshis::from_u64`.

Reply:
Yup, this is blocking.

Reply:
AFAICT this has not been resolved.